bootstrap/kubeadm/internal/cloudinit/kubeadm-bootstrap-script.sh

```diff
@@ -0,0 +1,117 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Log an error and exit.
+# Args:
+#   $1 Message to log with the error
+#   $2 The error code to return
+log::error_exit() {
+  local message="${1}"
+  local code="${2}"
+
+  log::error "${message}"
+  log::info "Removing member from cluster status"
+  kubeadm reset -f update-cluster-status || true
+  log::info "Removing etcd member"
+  kubeadm reset -f remove-etcd-member || true
+  log::info "Resetting kubeadm"
+  kubeadm reset -f || true
+  log::error "cluster.x-k8s.io kubeadm bootstrap script $0 exiting with status ${code}"
+  exit "${code}"
+}
+
+log::success_exit() {
+  log::info "cluster.x-k8s.io kubeadm bootstrap script $0 finished"
+  exit 0
+}
+
+# Log an error but keep going.
+log::error() {
+  local message="${1}"
+  timestamp=$(date --iso-8601=seconds)
+  echo "!!! [${timestamp}] ${1}" >&2
+  shift
+  for message; do
+    echo "    ${message}" >&2
+  done
+}
+
+# Print a status line.  Formatted to show up in a stream of output.
+log::info() {
+  timestamp=$(date --iso-8601=seconds)
+  echo "+++ [${timestamp}] ${1}"
+  shift
+  for message; do
+    echo "    ${message}"
+  done
+}
+
+check_kubeadm_command() {
+  local command="${1}"
+  local code="${2}"
+  case ${code} in
+  "0")
+    log::info "kubeadm reported successful execution for ${command}"
+    ;;
+  "1")
+    log::error "kubeadm reported failed action(s) for ${command}"
+    ;;
+  "2")
+    log::error "kubeadm reported preflight check error during ${command}"
+    ;;
+  "3")
+    log::error_exit "kubeadm reported validation error for ${command}"
+    ;;
+  *)
+    log::error "kubeadm reported unknown error ${code} for ${command}"
+    ;;
+  esac
+}
+
+function retry-command() {
+  n=0
+  local kubeadm_return
+  until [ $n -ge 5 ]; do
+    log::info "running '$*'"
+    # shellcheck disable=SC1083
+    "$@" --config /tmp/kubeadm-controlplane-join-config.yaml {{.KubeadmVerbosity}}
+    kubeadm_return=$?
+    check_kubeadm_command "'$*'" "${kubeadm_return}"
+    if [ ${kubeadm_return} -eq 0 ]; then
+      break
+    fi
+    # We allow preflight errors to pass
+    if [ ${kubeadm_return} -eq 2 ]; then
+      break
+    fi
+    n=$((n + 1))
+    sleep 15
+  done
+  if [ ${kubeadm_return} -ne 0 ]; then
+    log::error_exit "too many errors, exiting"
+  fi
+}
+
+retry-command kubeadm join phase preflight
+retry-command kubeadm join phase control-plane-prepare download-certs
+retry-command kubeadm join phase control-plane-prepare certs
+retry-command kubeadm join phase control-plane-prepare kubeconfig
+retry-command kubeadm join phase control-plane-prepare control-plane
+retry-command kubeadm join phase kubelet-start
+retry-command kubeadm join phase control-plane-join etcd
```

Comment:
If `kubeadm join phase control-plane-join etcd` don't we need to reset before retrying here ? retrying without reset would try to add a new member cc @vincepri @neolit123

Reply:
it is true that the phase adds a new member:
https://github.com/kubernetes/kubernetes/blob/91766b86a90cfaa472fbbb9b3c197e0f55e0200b/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go#L109

https://github.com/kubernetes/kubernetes/blob/91766b86a90cfaa472fbbb9b3c197e0f55e0200b/cmd/kubeadm/app/phases/etcd/local.go#L140